File: WalletKitJava/CoreNative/src/main/java/com/breadwallet/corenative/crypto/BRCryptoExportablePaperWallet.java
/*
 * Created by Ehsan Rezaie <ehsan@brd.com> on 11/23/20.
 * Copyright (c) 2020 Breadwinner AG.  All right reserved.
 *
 * See the LICENSE file at the project root for license information.
 * See the CONTRIBUTORS file at the project root for a list of contributors.
 */
package com.breadwallet.corenative.crypto;

import com.breadwallet.corenative.CryptoLibraryDirect;
import com.google.common.base.Optional;
import com.sun.jna.Native;
import com.sun.jna.Pointer;
import com.sun.jna.PointerType;

public class BRCryptoExportablePaperWallet extends PointerType {
    public BRCryptoExportablePaperWallet() { super(); }

    public BRCryptoExportablePaperWallet(Pointer address) {
        super(address);
    }

    public static BRCryptoExportablePaperWalletStatus validateSupported(BRCryptoNetwork network,
                                                                        BRCryptoCurrency currency) {
        return BRCryptoExportablePaperWalletStatus.fromCore(
                CryptoLibraryDirect.cryptoExportablePaperWalletValidateSupported(
                        network.getPointer(),
                        currency.getPointer()
                )
        );
    }

    public static BRCryptoExportablePaperWallet createAsBTC(BRCryptoNetwork network,
                                                            BRCryptoCurrency currency) {
        return new BRCryptoExportablePaperWallet(
                CryptoLibraryDirect.cryptoExportablePaperWalletCreateAsBTC(
                        network.getPointer(),
                        currency.getPointer()
                )
        );
    }

    public Optional<BRCryptoKey> getKey() {
        Pointer thisPtr = this.getPointer();

        return Optional.fromNullable(
                CryptoLibraryDirect.cryptoExportablePaperWalletGetKey(thisPtr)
        ).transform(BRCryptoKey::new);
    }

    public Optional<BRCryptoAddress> getAddress() {
        Pointer thisPtr = this.getPointer();

        return Optional.fromNullable(
                CryptoLibraryDirect.cryptoExportablePaperWalletGetAddress(thisPtr)
        ).transform(BRCryptoAddress::new);
    }

    public void give() {
        Pointer thisPtr = this.getPointer();

        CryptoLibraryDirect.cryptoExportablePaperWalletRelease(thisPtr);
    }
}

File: WalletKitCore/src/crypto/BRCryptoExportablePaperWallet.c
//
//  BRCryptoExportablePaperWallet.c
//  BRCore
//
//  Created by Ed Gamble on 2/23/21
//  Copyright © 2020 breadwallet. All rights reserved.
//
//  See the LICENSE file at the project root for license information.
//  See the CONTRIBUTORS file at the project root for a list of contributors.

#include "BRCryptoExportablePaperWalletP.h"
#include "BRCryptoHandlersP.h"

private_extern BRCryptoExportablePaperWallet
cryptoExportablePaperWalletAllocAndInit (size_t sizeInBytes,
                                         BRCryptoBlockChainType type,
                                         BRCryptoNetwork network,
                                         BRCryptoAddress address,
                                         BRCryptoKey key) {
    assert (sizeInBytes >= sizeof (struct BRCryptoExportablePaperWalletRecord));
    assert (cryptoKeyHasSecret (key));

    BRCryptoExportablePaperWallet wallet = calloc (1, sizeInBytes);

    wallet->type = type;
    wallet->sizeInBytes = sizeInBytes;
    wallet->handlers = cryptoHandlersLookup(type)->exportablePaperWallet;

    wallet->network  = cryptoNetworkTake (network);
    wallet->address  = cryptoAddressTake (address);
    wallet->key      = cryptoKeyTake     (key);

    return wallet;
}

private_extern void
cryptoExportablePaperWalletReleaseInternal (BRCryptoExportablePaperWallet wallet) {
    cryptoNetworkGive (wallet->network);
    cryptoAddressGive (wallet->address);
    cryptoKeyGive     (wallet->key);

    memset (wallet, 0, wallet->sizeInBytes);
    free (wallet);
}

extern BRCryptoExportablePaperWalletStatus
cryptoExportablePaperWalletValidateSupported (BRCryptoNetwork network,
                                              BRCryptoCurrency currency) {
    if (CRYPTO_FALSE == cryptoNetworkHasCurrency (network, currency))
        return CRYPTO_EXPORTABLE_PAPER_WALLET_INVALID_ARGUMENTS;

    const BRCryptoHandlers *handlers = cryptoHandlersLookup (network->type);

    return ((NULL != handlers->exportablePaperWallet &&
             NULL != handlers->exportablePaperWallet->validateSupported)
            ? handlers->exportablePaperWallet->validateSupported (network, currency)
            : CRYPTO_EXPORTABLE_PAPER_WALLET_ILLEGAL_OPERATION);
}

extern BRCryptoExportablePaperWallet
cryptoExportablePaperWalletCreate (BRCryptoNetwork network,
                                   BRCryptoCurrency currency) {
    const BRCryptoHandlers *handlers = cryptoHandlersLookup (network->type);

    return ((NULL != handlers->exportablePaperWallet &&
             NULL != handlers->exportablePaperWallet->create)
            ? handlers->exportablePaperWallet->create (network, currency)
            : NULL);
}

extern void
cryptoExportablePaperWalletRelease (BRCryptoExportablePaperWallet wallet) {
    if (NULL != wallet->handlers->release)
        wallet->handlers->release (wallet);
    else
        cryptoExportablePaperWalletReleaseInternal (wallet);
}

extern BRCryptoKey
cryptoExportablePaperWalletGetKey (BRCryptoExportablePaperWallet wallet) {
    return cryptoKeyTake (wallet->key);
}

extern BRCryptoAddress
cryptoExportablePaperWalletGetAddress (BRCryptoExportablePaperWallet wallet) {
    return cryptoAddressTake (wallet->address);
}
